sql-parser: support general typed string literals

benesch opened the pull request:
Support PostgreSQL's typed string literals, as in:

    SELECT bool 'true'

Previously we only supported the typed string literals for the date,
timestamp, and timestamp with time zone types as required by the SQL
standard, but PostgreSQL generalizes this syntax for all types.

As a side effect, this fixes the bug where

    SELECT date

parsed as a syntax error rather than as selecting a column named "date".
The fix also generalizes, e.g., `SELECT bool` is also valid.

Fix #2577.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/materializeinc/materialize/3146)
<!-- Reviewable:end -->

benesch commented:
Note that this PR introduces some regressions, because it moves some validation that used to happen at plan time to runtime. I'll open a separate PR to re-introduce those errors as runtime errors. There's also possibly a missing optimization.

justinj commented on src/sql-parser/src/ast/mod.rs:
```diff
@@ -258,6 +261,10 @@ pub enum Expr {
     Nested(Box<Expr>),
     /// A literal value, such as string, number, date or NULL
     Value(Value),
+    TypedString {
```
No strong feelings from me but seems like we could also just parse this as a cast and avoid adding a new AST node?

benesch replied:
Yeah I went back and forth on that a lot. Definitely can parse as a cast, but it seemed like maybe `SHOW CREATE VIEW` would want to preserve `SELECT date '2010-01-01'` rather than rewriting to `SELECT CAST ('2010-01-01' as date)`. Hmm, maybe we could instead rewrite to `SELECT '2010-01-01'::date`. What do you think?

justinj replied:
I like `'2010-01-01'::date`! In general I'm sympathetic to both lowering the number of AST nodes and to not mangling the input too much at the parsing phase, so I'm happy either way.

benesch replied:
👍 went with the cast now that the output for casts is more succinct.

justinj commented on src/sql-parser/src/parser.rs:
```diff
@@ -43,6 +43,14 @@ macro_rules! parser_err {
     };
 }
 
+macro_rules! maybe {
```
Is this a standard macro? If not, consider fusing it with `maybe_parse` since they always seem to be called together?

benesch replied:
Nope, just made it up. It's sort of the inverse of the `try!`/`?` macro—early return if you get the success case rather than the error case.

I really wanted to figure out how to get a `maybe_parse!` macro to work, but I couldn't come up with a design that seemed quite right because of macro hygiene issues. The macro needs to take `self` as input in order to have access to it so it reads kind of weird and is super specific:

```
maybe_parse!(self, |parser| {
    // ...
});
```

But maybe that's preferable! Not sure! Thoughts?

justinj replied:
Ah I see! I didn't know about the hygiene issues—no strong feelings from me!

benesch commented:
Ok, addressed all your feedback, except there is a constant expr that is not getting reduced: `strtodate("1995-01-01")`

Digging in now. Super suspicious because we're able to reduce these constant expressions everywhere else.

benesch commented:
Oh, dur, that plan is an `EXPLAIN RAW PLAN`. Everything here is working as intended, apologies.
